Color picker

bjarnef opened the pull request:
Follow-up on https://github.com/umbraco/Umbraco.UI/pull/112

## Description

This PR add a Color Picker components splitted into smaller separate components. Furthermore it introduces a few utility functions like `clamp`.

```
// Clamp number between two values with the following line:
const clamp = (num, min, max) => Math.min(Math.max(num, min), max);
```

An alternatively version of `clamp` function can be found here: https://www.webtips.dev/webtips/javascript/how-to-clamp-numbers-in-javascript

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [x] New feature (non-breaking change which adds functionality)

## Motivation and context

<!--- Why is this change required? What problem does it solve? -->

## How to test?

## Screenshots (if appropriate)

https://user-images.githubusercontent.com/2919859/183895436-d6d77811-d376-40e4-905b-9fe86976e592.mp4

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply.  If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] If my change requires a change to the documentation, I have updated the documentation in this pull request.
- [x] I have read the **[CONTRIBUTING](<(https://github.com/umbraco/Umbraco.UI/blob/dev/docs/CONTRIBUTING.md)>)** document.
- [ ] I have added tests to cover my changes.

## TODO
- [ ] Make vertical slider go from min at bottom to max at top.
- [ ] Update swatches with displayValue or label.
- [x] Update handle position in color area when selecting color from palette.
- [x] Ensure background color change in opacity slider. Maybe we need to separate into `uui-color-hue-slider` and `uui-color-opacity` to make the markup a bit different as the opacity slider need a layer for transparency and a layer with background color. These could eventually share a `ColorSliderMixin`.
- [x] Can we configure color picker as inline vs. popup, e.g. using popover component if not inline?
- [x] Support transparent colors in swatches, e.g. rendered with checkered background?
- [ ] Deselect color in palette, when color is changed via color area or via hue slider.
- [x] Focus handle inside color area and color sliders when navigating using keyboard.
- [x] Close popover using esc key.
- [x] Color is copied when clicking the preview button, but should this has any focus/animation like in the Shoelace library? Furthermore it seems the inner button element doesn't fill full height of `uui-button` which is seen with a transparent color on hover/focus.
- [x] ~~Can we use `aria-label` without a button text? It seems setting `label` on `uui-button` both set `aria-label` and button text.~~ using button element instead.

bjarnef commented:
@iOvergaard @nielslyngsoe @madsrasmussen I decided to go with the Colord library as it seems most flexible yet lightweight: https://colord.omgovich.ru/

iOvergaard commented:
Very good, @bjarnef !

bjarnef commented:
@iOvergaard a bit more progress. In the `uui-color-area` I need to set color or hue value to update the spectrum as the hue value change.

```
<uui-color-area
          @change=${this.handleGridDrag}
          >
</uui-color-area>
```

Currently the color spectrum is generated based on this inside `uui-color-area`.

```
style=${styleMap({ backgroundColor: `hsl(${this.hue}deg, 100%, 50%)` })}
```

I tried something like this, but how would I get the updated value inside the `uui-color-area` as the hue/color change? Maybe using `watch` on the `value` property?

```
<uui-color-area
          .value=${live(this.value)}
          @change=${this.handleGridDrag}
          >
</uui-color-area>
```

bjarnef commented:
@iOvergaard could you help a bit with my questions in https://github.com/umbraco/Umbraco.UI/pull/280#issuecomment-1210579049

iOvergaard commented:
Hi @bjarnef 

Sorry, yes. If you want to react on the property called "value", you could do so by making a getter/setter like `get value()` and a `set value(newVal)` function, or you could react to changes in the lifecycle [update](https://lit.dev/docs/v1/components/lifecycle/#update) or updated callbacks. Does that help you in any way?

If you want to return an object like `colord` you could do good by embedding it into an object that we control, so that if we ever switch colord with something else in the future, we won't have to change the entire event structure.

bjarnef commented:
Okay, I did try something like that, but didn't quite get it to work. Maybe I missed something, so I will have another look.

Would you prefer the event return object with an instance of Colord or the specific properties like r, g, b etc.? Do you have a example of another component returning an object when an event occurs?

iOvergaard commented:
@bjarnef The combobox for instance returns a [UUIComboboxEvent](https://github.com/umbraco/Umbraco.UI/blob/v1/dev/packages/uui-combobox/lib/UUIComboboxEvent.ts) with a few string properties taken from an internal change event. There are other examples as well through out the library. I don't know exactly what properties are useful to have from `colord` but it sounds like a good approach to pull out details from their object, so that we can more easily swap it in the future if the need arises.

bjarnef commented:
@iOvergaard it kinda works now, but not sure about the value property as when color is `#fff` hue will be 0.

https://user-images.githubusercontent.com/2919859/184881680-7dd3ddc4-8c07-4d50-b193-2ea9739e78ea.mp4

I don't think we want to change current hue value in this case, in this example the hue should still be 116.

![image](https://user-images.githubusercontent.com/2919859/184883031-62ba6659-92ed-47ae-a901-46dae723f5e6.png)

iOvergaard commented:
Not sure what you mean exactly, @bjarnef. The Selectable mixin just listens to click events and then sends either a SELECTED or UNSELECTED event bubbling up through the stack so you'll need something high up in the stack to catch the events that will then forward the appropriate values down to the other components.

bjarnef commented:
@iOvergaard I was looking at the example here:
https://github.com/umbraco/Umbraco.UI/blob/v1/dev/packages/uui-ref/lib/uui-ref.element.ts#L19-L21

Besides the styling of the selected attribute I am not sure where I get the value related to selected element. Furthermore can we control only at single element to be selected at the time? E.g. the color of the selected swatch. In the color picker we may not neccessary style the selected color swatch, but could be useful if it is used at icon picker in backoffice, if we want this part like it is today.

iOvergaard commented:
@bjarnef Uhh the value of the selected element. It's not included in the custom event, however you should be able to find it on the composed path if you are not listening to each color individually: https://developer.mozilla.org/en-US/docs/Web/API/Event/composedPath

Does that help you in any way? Otherwise I might be able to get someone else to help you next week ☺️

bjarnef commented:
@iOvergaard I had a look at the `uui-combobox` and `uui-combobox-list` components. From what I can see these are currently only single pickers? I guess it would make sense it these both support single/multi values.

We probably need something similar in color swatches. In Color Picker it would most likely be a single selection, but there may be other use-cases where multi selection makes sense.

https://user-images.githubusercontent.com/2919859/185665818-2e5ba1d7-9bb1-46f4-99b2-8fe6bd349f1e.mp4

bjarnef commented:
https://github.com/umbraco/Umbraco.UI/pull/280/files/211f00a80d05433d0147c1a06190a2b2d52935a9..e206638269b17393138d0e2eabbc50ce520b3877 makes the convertion works from color names like, `tomato`, `skyblue`, `lavendel` etc.

bjarnef commented:
@iOvergaard @nielslyngsoe I have updated this PR with a few things that needs to be done, but in general it's getting there 😎

It would be fantastic if some of you could try this out 🤞🙌

iOvergaard commented:
Hi @bjarnef 
We are planning to review this very soon, and as you pointed out it would be useful in the new backoffice.

loivsen commented:
ColorArea
Default color seems to be white (or perhaps none), however the picker shows in the red area on load.
Preferably they should be the same/match from the beginning if possible

ColorPicker
The color sliders changes the width depends on the text of the color format button
The eyedropper button sometimes open previous windows tab (thus tabbing out of the browser)
Mouseover on the color preview shows a square when there is transparency (is this on purpose?)
Swatches
I miss some kind of indication for "this color is chosen". Forexample a checkmark on the chosen swatch color.
Idea: It could also be a setting you could turn on and off for the swatches if not everybody would want this..

bjarnef commented:
@loivsen yes, the default color is empty, but it is difficult to select an empty color in the color area.
https://shoelace.style/components/color-picker

but maybe empty value should show transparent (checkered background) and no selection in color area like this current color picker in Umbraco: https://seballot.github.io/spectrum/

The format button should probably have a fixed or minimum width to ensure it doesn't change as format/text is changed (unless a monospaced font was used and same number of characters).

I haven't noticed the eyedropper open/changing window/tab, but it does allow to pick any color on screen - also outside browser window.

Regarding swatches the selected color currently has a `selected` attribute, so we could easily style this with a checkbox, but we need to update this it color later is changed to another color via input, sliders or color spectrum. In this case the (and if used next to icon picker as today) it would probably be a single picker, but if used later by package developer it might make sense to allow picking multiple colors.

Currently the swatches should only render if any swatches has been provided, so I guess it's a setting to disable the swatches, but it could be a separate setting.

iOvergaard commented:
Hi @bjarnef, is there anything we can assist you with from HQ's side?

bjarnef commented:
@iOvergaard I noted a few things here https://github.com/umbraco/Umbraco.UI/pull/280#issue-1327540098 which would be great to look at.

I think we should consider if the active color swatches should be selected in color picker and configure the color swatches to allow single/multiple selections.

Furthermore as @loivsen mentioned https://github.com/umbraco/Umbraco.UI/pull/280#issuecomment-1250683770 when a transparent color is selected (the preview styling on `uui-button` looks a bit strange on focus).

We also need to change color in opacity slider as hue change, but I am not sure if the inner element in can be styled (e.g. via a property). Alternatively we may need to split these into two components (but it may be useful with a base ColorSlider mixin / component to inherit from, e.g. to get the basic functionality with min, max etc).

Maybe a empty color shouldn't show a selected color in the color area, but some color pickers tends to use "red" as default selection in the HSB-grid.

bjarnef commented:
@iOvergaard I have merged with latest changes, but I now get some errors regarding import of the utils here:

![image](https://user-images.githubusercontent.com/2919859/195193373-4267a835-82a8-439d-aca4-9ec3ec44f6d8.png)

bjarnef commented:
@iOvergaard do you know what is cause the issue here? Furthermore I noticed there is no a `v1/contrib` branch... should I target the PR that branch instead?

iOvergaard commented:
Trying to look into it, @bjarnef. It's a bit odd. Seems like it doesn't register any of the new types unless explicitly building the uui-base package. But even then it complains.

iOvergaard commented:
You can continue targeting v1/dev here. We want to use v1/contrib moving forward for external pull requests.

bjarnef commented:
@iOvergaard yes I could access e.g. `demandCustomElement` and `Timer` from `uui-color-slider` because these existed in `node_modules` folder here:
```
umbraco-ui\packages\uui-color-slider\node_modules\@umbraco-ui\uui-base\lib\utils

iOvergaard commented:
I figured it out, I think. There were some references to an older version of uui-base which resulted in the package being symlinked incorrectly. 

I allowed myself to push to your branch - hope that's okay!

iOvergaard commented:
> But clamp and math wasn't copied to this folder (I guess it failed before building/copying these).

Yeah, we are using npm workspaces here so it actually symlinks the folder `./node_modules/@umbraco-ui/uui-base` to `./packages/uui-base` but I guess it doesn't work well if you target another version than what uui-base currently claims to be (in this case 1.0.0-rc.3 --> 1.0.0). 

Normally lerna will keep these up-to-date, but I guess since your library only exists in your fork, the lerna commands were never run in your copy when we released 1.0.0.

bjarnef commented:
@iOvergaard yes, it seems to work after your change in https://github.com/umbraco/Umbraco.UI/commit/18097f1a58eb56338c1b5c31357c48ec83c3eeaf

I thought I tried that, but maybe something else failed at that time.

bjarnef commented:
@iOvergaard one issue I have it when styling `uui-button` I get this on hover/focus:

![image](https://user-images.githubusercontent.com/2919859/195874082-f98f72ef-cf24-4606-8f3d-51a361feafbd.png)

![image](https://user-images.githubusercontent.com/2919859/195874144-b8ee7703-624f-4b7b-a823-470bf8abadb7.png)


Furthermore I have an `inline` mode using `uui-popover`, but it doesn't seem to work well with `uui-color-picker` inside it.
Sometimes it also "jumps" a bit to the correct position.

https://user-images.githubusercontent.com/2919859/195874809-95de1add-b57f-4a1a-a582-5123aae1a468.mp4

bjarnef commented:
@iOvergaard I worked a bit more on this, so now color swatches accept transparency colors as well (the checkered style is currently used a few places, so maybe it could be added globally somewhere).

![image](https://user-images.githubusercontent.com/2919859/195991709-08c4d8cf-728a-40bb-a119-d8095c655d2c.png)

The color picker use the `uui-popover`, but unlike the docs here https://uui.umbraco.com/?path=/docs/uui-popover--aaa-overview  , I have moved the trigger outside the `uui-popover` element (otherwise the popover isn't closed when clicking the area to the right for trigger button).
I works great (although the popover seems to "jump" a bit to correct position) and when viewing in docs it opens behind the code sample.

Furthermore I am not sure how the hover on `uui-button` (I could adjust the padding inside overwriting a few variables though).

https://user-images.githubusercontent.com/2919859/195991778-da2416c9-ef73-4df2-8aa3-00da5ed47496.mp4

bjarnef commented:
I also noticed `uui-input` had a method `focus()`, but not a `blur()` or `select()` like the Shoelace library had in their "input" component, e.g. when the preview color is clicked it select and copy the value from input field.

bjarnef commented:
@iOvergaard @nielslyngsoe @madsrasmussen I think this is close to being ready to review, but there a few more todos I would like to finish: https://github.com/umbraco/Umbraco.UI/pull/280#issue-1327540098 and of course cleanup 

But feel free to test it further, check naming convention, code style etc. 👍

bjarnef commented:
@iOvergaard not sure if this is best practice, but it works: https://github.com/umbraco/Umbraco.UI/pull/280/commits/d62a56fd1dc4bddcb2055cda2b4bb3a692d448da

Could maybe be separated into two components for easier use.

https://user-images.githubusercontent.com/2919859/196813039-c315aa64-26d1-4f33-8234-a975a601674b.mp4

iOvergaard commented:
Hi @bjarnef 

Looks good so far! 

We've put in a review task for your PR in an upcoming sprint at the end of this month by @nielslyngsoe - hope that will be useful to you :-)

nielslyngsoe commented on packages/uui-color-slider/lib/uui-color-slider.element.ts:
```diff
@@ -0,0 +1,291 @@
+      this.isVertical = true;
+    } else {
+      this.isVertical = false;
+    }
```
Forget about this feedback, Its better to go with the vertical attribute that I proposed above.

Then this would go away :-)

nielslyngsoe reviewed (commented):
Hi @bjarnef 

I've been going through this code, I really love it.

I wanted to get this code up to the same standard as the rest of our code(at least towards where we aim to be) and use this opportunity to give you our own learnings.

I haven't been able to get through it all, but I think you will learn a few new tricks and be able to implement this across.

Let me know any thoughts and concerns, as well I would like to learn how this approach seems for you. As I like to be able to write some guidelines on our architecture at one point.

I also have some design feedback, which is part of my next comment :-)

nielslyngsoe commented:
Hi @bjarnef 

I finally have some time to look at this wonderful PR. Let me try to get around to everything though I properly will miss something.

First I have to say I love this, it seems to work very well and I think you have organized it well :-D

**1. using the uui-button for the color-preview:**
 I'm not totally sure here, but I think it only makes sense to use the `uui-button` if its going to look like the other buttons.
 In this case the color-picker-preview is sortof its own thing, so I think its would make sense to make it its own component and not extending the uui-button. Then these troubles should be gone.
 
**2. Display which color swatch is selected
To be consistent with the 'new' way we do this, I would recommend adding a blue border around the picked color-swatch. 
See some of the Cards for reference. This one might need to have a black or dark-transparent 1px border on the inside in between the actual color and the select-border. For visual distinguishment. Note how this plays with the focus-outline.

3**. Styling the buttons next to the color-string input**
Again on the very detail-oriented side of visuals, it would be nice to display the buttons adjacent to the color-string-input in this way, so they get visual relation.
![image](https://user-images.githubusercontent.com/6791648/203326631-494f88ec-820d-4868-9dd4-d835dc216638.png)
This I just made by removing the look from the buttons. Adding a border to the button group but not a left border.
Ensuring uui-buttons has --uui-button-height set to 100% and the second uui-button gets a left border.
(not on this screenshot, but it would also be great to make the text on the input-format-switcher go uppercase `text-transform: uppercase`

**4. Sliders could be more minimalistic, to make a calm layout.**
I would hide the display of the value and make the slider-bar just 6px heigh. To bring more focus to the picked color and make a more calm overall look:
![image](https://user-images.githubusercontent.com/6791648/203330108-1cc942cb-b80e-4687-b4c9-c3cbd2978192.png)
BTW. great that the slider-handles are similar to the color-picker target, this gives a nice harmony!

**5. Copying the color on click**
Just a great detail, thanks!

**6. There is a disconnection between color-swatches and the rest.**
It goes both ways. If i pick a swatch and then start manipulating it, then the swatch should be de-selected.
And when I Pick a swatch the color-picker target is not updated.
Let me know if there is a concrete coding problem with this, then I would like to help out :-)

**7. The color slider first rounds its value on pointer-up.**
I think it's best to have the value rounded/corrected on pointer-move as the user then would get the same value while dragging as when releasing the cursor.  And I noticed currently the pointer-up logic only fires if the user releases the cursor on top of the handle. Maybe this would be fixed when implementing the onStop callback from `drag`. :-)


Okay, that was a lot, sorry if you get overwhelmed. Overall these are the final details. let me know if you like me to do some of the implementations.

Thanks in advance 💙

bjarnef commented:
@nielslyngsoe I have fixed a few of the issues, but it seems I get some import errors after pulling latest changes into this branch,

![image](https://user-images.githubusercontent.com/2919859/203402857-091e27bc-7ec7-444d-91c4-f2a24cbfb08f.png)

The values at the sliders shouldn't be there, but it was mainly to output/debug the value and the rounded value.

nielslyngsoe commented:
Hi, @bjarnef sounds good, and sorry for the amount of feedback. I wanted to make sure you got everything :-)

Could you try running `npm install`  I think we might have an `issue` when new components come around, then you would need to re-install for the references to be updated. I'm thinking this could be the case as I cannot replicate the issue here. :-/

Let me know if that fixes it. I guess there is a potential for some NPM script improvements here.

bjarnef commented:
It is fine :) I also had a lot of things I wanted to do and much of it has been solved.

I tried with `npm install` and I could run storybook, but it then seems to fail somewhere importing the other components. I think it is because they wasn't copied as node modules inside the package/component folder.

I think it was a similar issue which @iOvergaard looked at, but at that time it was a mismatch with package versions I think.

iOvergaard commented:
Hi @bjarnef 

I updated the dependencies of the library, particularly there was an update on the main branch to the way rollup is used. I don't know if that is what you ran in to. It also seems you are using an older version of NPM seeing that the lockfile was quite old. Try and use Node 16 / NPM 8.15+, if you aren't already, and see if it works now with the latest updates.

I also fixed some inter-package dependencies we would have ran into, if we tried using the packages individually, namely dependencies between uui-color-swatches and uui-color-swatch 👍

bjarnef commented on packages/uui-color-area/package.json:
```diff
@@ -30,7 +30,8 @@
   "dependencies": {
-    "@umbraco-ui/uui-base": "1.0.0"
+    "@umbraco-ui/uui-base": "1.0.0",
+    "colord": "^2.9.3"
```
@iOvergaard should this be specific on version `2.9.3` as well?

iOvergaard commented:
BTW build is failing because of eslint issues.

bjarnef commented:
Yeah, command says these versions of NodeJS and NPM, but pretty sure I updated these last time 😁

![image](https://user-images.githubusercontent.com/2919859/203549371-e2dc15fc-7e17-46fb-8ba1-03de00b28b95.png)

iOvergaard commented:
> Yeah, command says these versions of NodeJS and NPM, but pretty sure I updated these last time 😁

Haha yes, the entire package-lock format changed between node 14 and 16, so try with that. Haven't tested with the latest LTS (node 18) yet, though, but that might also be worth a try.

bjarnef commented:
@iOvergaard I upgraded NPM to v8 and NodeJS to v18.

![image](https://user-images.githubusercontent.com/2919859/203864221-15a776af-4023-44f6-a6e8-82139a6dfd30.png)

However I get this error using NodeJS v18.12.1

![image](https://user-images.githubusercontent.com/2919859/203864264-5bd586ed-45e8-4bfe-8464-ae19c831ce6f.png)

bjarnef commented:
@iOvergaard @nielslyngsoe I got it working again uninstalling NodeJS 18 and installing latest NodeJS 16.

Do we have any special animations in the UI library?
Currently I have this when copying the color from the preview button.

https://user-images.githubusercontent.com/2919859/203930978-84ac3310-d3b0-4306-845e-a6ee6cc3edb6.mp4

bjarnef commented:
@nielslyngsoe regarding card selection / focus I noticed on click in looks like this.

![image](https://user-images.githubusercontent.com/2919859/203934927-3e1b2dfa-ccc4-4846-9139-3ea39fd4a3ce.png)

but focus using keyboard it looks like this. Is that expected?

![image](https://user-images.githubusercontent.com/2919859/203934846-c798053c-caa7-42d0-b41b-ffb93af59402.png)

iOvergaard commented:
> @iOvergaard I upgraded NPM to v8 and NodeJS to v18.

Yeah, I suspected as much. Expectedly some breaking changes have been introduced and we need to update all our products to node 16. We usually mark what is supported in the package.json:

https://github.com/umbraco/Umbraco.UI/blob/3dcb9ce2abd59dbb665ca0391aae9b72afa930f8/package.json#L53:L56

bjarnef commented:
Hi @nielslyngsoe 

I have modified this accordion to your suggestions:

**1. using the uui-button for the color-preview**
- The preview button is a regular `button` element instead of using `uui-button`. it is similar button used as button (when not inline)

**2. Display which color swatch is selected**
It is using same styling as it could find in `uui-card`. I have changes to outline border to 1px though.

3**. Styling the buttons next to the color-string input**
I have styled a bit on the buttons here. I think we can't transforms text in `uui-button` directly without uppercase the actual text or in these case I wrapped the text in a nested span in the slot to transform text to uppercase.

4. Sliders could be more minimalistic, to make a calm layout.
I haven't changed this yet, but I think it would be great to still see the color spectrum and alpha gradient, so I didn't want to make the height too small and as it is also possible to click/touch the area, but 10px instead of the current 15px may be better?

**5. Copying the color on click**
I think the pulse animation should be moved somewhere to re-use. Not sure which color should be used here in the effect. It could use the current preview color, but not sure that would work with a very light color.

**6. There is a disconnection between color-swatches and the rest.**
I am not sure how to update this between the components. I could use a bit help with this :) 
The uui-color-swatches has a change event, but not sure how we could update the selected state of the nested uui-swatch.
The swatches are currently a string array, but maybe it should also be an option with Array<Item> so each item e.g. both has color value and name/tooltip. E.g. if we intend to use swatches in other contexts with a name/label.

**7. The color slider first rounds its value on pointer-up.**
Currently it seems to work well, but I noticed a few cases where it seems to round the value to a different number from when I stopped dragging and releasing handle.


Furthermore I have also added the `vertical` attribute to `uui-color-slider` (in vertical mode it think we need to revert the min/max value, so low value it at bottom and high value is at top. I tried with CSS transform, but could make that work.

I also fixed where color/handle wasn't updated in `uui-color-area` which selecting a swatch color and initial color set from value property.

https://user-images.githubusercontent.com/2919859/204136935-2e728c0e-cd0b-4a5d-980d-2099d301e75d.mp4

iOvergaard commented:
Hi @bjarnef and happy new year!

I have synced your branch with v1/contrib and tried to clean-up a bit to make the build succeed (it was failing because of eslint rules). While I was at it, I did some more stuff to look into the last few bits to make it all come together:

1. I added support for actions in Storybook on most components
2. I moved the uui-color-* packages into a Color category in Storybook
3. I added the uui-color-* dependencies to the meta UUI package to enable the builder to catch and release them
4. I added a CHANGE event to uui-color-picker (seems like a good idea to know when a color has been picked?)
5. uui-color-slider: added support for ArrowUp and ArrowDown (especially useful on vertical=true)
6. Removed a few console logs and also a raw `Value: ${value}` on swatches

We have been using a custom build of uui-color-swatches and uui-color-swatch on the new backoffice for a while now, and they are looking good.

I want to see if I can get someone to review this very soon, so we can get it released. I'm thinking to do an RC of UUI 1.1.0 perhaps and include these new components then.

bjarnef commented:
Hi @iOvergaard 
Sounds great and happy new year to you as well 🎉

Regarding vertical color slider I also think we need a change to make it go from low value at bottom and high value at top.

Furthermore I guess we need to sync the colors so it isn't selected (unselected) in e.g. color swatches if changes elsewhere and the opposite if color is changed to a matching color in the swatches?

iOvergaard commented:
@bjarnef The vertical slider and value/label are listed as TODOs still in the description of this PR, right?

Regarding syncing of colors, I guess the uui-color-picker component should pick up the events from sub-components and update the others when it detects a change?

bjarnef commented:
@iOvergaard yes here :) 
https://github.com/umbraco/Umbraco.UI/pull/280#issue-1327540098

Yes, I guess that it is how it should work.

Currently the color swatches are based on string array of colors, but could it also accept Array<T> where T contains label and value. E.g. if we want to show tooltip/label for each color swatch. Maybe we also want something similar to the advanced color swatches with labels in the current approved color picker in Umbraco. These could probably use the card component I guess?
